misc/python/materialize/cli/mzcompose.py

```diff
@@ -149,6 +169,24 @@ def gen_shortcuts(repo: mzbuild.Repository) -> int:
     return 0
 
 
+def list_ports(composition: mzcompose.Composition, service: str) -> List[int]:
+    """Find all ports open on the host for a given service"""
+    # Parsing the output of `docker-compose ps` directly is fraught, as the
+    # output depends on terminal width (!). Using the `-q` flag is safe,
+    # however, and we can pipe the container IDs into `docker inspect`,
+    # which supports machine-readable output.
+    containers = composition.run(["ps", "-q"], capture=True).stdout.splitlines()
+    metadata = spawn.capture(["docker", "inspect", "-f", "{{json .}}", *containers,])
+    metadata = [json.loads(line) for line in metadata.splitlines()]
+    ports = []
+    for md in metadata:
+        if md["Config"]["Labels"]["com.docker.compose.service"] == service:
+            for (name, port_entry) in md["NetworkSettings"]["Ports"].items():
+                for p in port_entry or []:
+                    ports.append(p["HostPort"])
+    return ports
```

Comment:
This seems like it wants to be a method on `Composition`.

Reply:
Can do!

Reply:
Turns out it already was!